### PR TITLE
Use extern aliases in BuildValidator

### LINKED ISSUE #51523: PDBs should include information about extern aliases
Several Roslyn projects don't rebuild with diagnostics like the following:

`C:\Users\rikki\src\roslyn\src\Test\PdbUtilities\Reader\CustomDebugInfoUtilities.cs(7,14): error CS0430: The extern alias 'DSR' was not specified in a /reference option`

It looks like extern aliases aren't included in the PDB compilation options. This should be fixed for 16.10.

### PR DESCRIPTION
**Pull request by RikkiGibson**
Closes #51523

### CODE REVIEW

File: eng/test-rebuild.ps1 [hunk 1]
```diff
@@ -37,55 +37,89 @@ try {
...
   " --assembliesPath `"$ArtifactsDir/obj/Microsoft.CodeAnalysis.Remote.Razor.ServiceHub`"" +
+  " --assembliesPath `"$ArtifactsDir/obj/Microsoft.CodeAnalysis.Remote.ServiceHub`"" +
   " --assembliesPath `"$ArtifactsDir/obj/Microsoft.CodeAnalysis.Remote.Workspaces`"" +
   " --assembliesPath `"$ArtifactsDir/obj/Microsoft.CodeAnalysis.ResultProvider.Utilities`"" +
   " --assembliesPath `"$ArtifactsDir/obj/Microsoft.CodeAnalysis.ResultProvider`"" +
   " --assembliesPath `"$ArtifactsDir/obj/Microsoft.CodeAnalysis.Scripting.TestUtilities`"" +
   " --assembliesPath `"$ArtifactsDir/obj/Microsoft.CodeAnalysis.Scripting`"" +
   " --assembliesPath `"$ArtifactsDir/obj/Microsoft.CodeAnalysis.Test.Utilities`"" +
   " --assembliesPath `"$ArtifactsDir/obj/Microsoft.CodeAnalysis.TestSourceGenerator`"" +
+  " --assembliesPath `"$ArtifactsDir/obj/Microsoft.CodeAnalysis.VisualBasic.CodeStyle.Fixes`"" +
```
Comment (RikkiGibson) [inline eng/test-rebuild.ps1:82]: I imagine many of these newly compatible projects were actually fixed by Jared's VB changes.

**Review by jaredpar — APPROVED** [review]

**Comment by RikkiGibson** [comment]
Tracking down the rebuild failure here. It seems like the netframework reference assemblies on the build agent are incomplete?

CI log:
```log
2021-03-01T21:27:56.1028634Z Verifying D:\workspace\_work\1\s\artifacts\obj\Microsoft.CodeAnalysis.VisualBasic.EditorFeatures.UnitTests\Release\net472\Microsoft.CodeAnalysis.VisualBasic.EditorFeatures.UnitTests.dll with pdb [embedded]
2021-03-01T21:27:56.1033277Z   Encoding
2021-03-01T21:27:56.1033677Z     default-encoding - 
2021-03-01T21:27:56.1034043Z     fallback-encoding - 
2021-03-01T21:27:56.1070906Z   Locating metadata references
2021-03-01T21:27:56.8176312Z Caching [01fd1641-bebf-46e1-9111-b588fbab11b2, C:\Users\vsagent\.nuget\packages\microsoft.netframework.referenceassemblies.net472\1.0.0\build\.NETFramework\v4.7.2\System.Windows.Controls.Ribbon.dll]
2021-03-01T21:27:56.8183096Z Caching [1cdb3037-702b-491a-b91c-b5f147960f2d, C:\Users\vsagent\.nuget\packages\microsoft.netframework.referenceassemblies.net472\1.0.0\build\.NETFramework\v4.7.2\UIAutomationClientsideProviders.dll]
2021-03-01T21:27:57.4769186Z Caching [76d78e3b-8071-477b-a3a1-57725b22b48d, D:\workspace\_work\1\s\artifacts\bin\Microsoft.CodeAnalysis.EditorFeatures.DiagnosticsTests.Utilities\Release\net472\ref\Microsoft.CodeAnalysis.EditorFeatures.DiagnosticsTests.Utilities.dll]
2021-03-01T21:27:57.9436699Z Missing metadata references:
2021-03-01T21:27:57.9437483Z   Microsoft.VisualBasic.dll - 02d1b628-df99-421c-a9d8-fd869aa945c1
2021-03-01T21:27:57.9440039Z Cannot resolve 1 references
```

Local run includes this resolved metadata reference:
```log
"C:\Users\rikki\.nuget\packages\microsoft.netframework.referenceassemblies.net472\1.0.0\build\.NETFramework\v4.7.2\Microsoft.VisualBasic.dll" - 113d973b-2b75-477a-b369-bfd9606423fc
```

**Comment by RikkiGibson** [comment]
I found that the assembly is missing even when we can verify that a local nuget cache is in use. I really have no idea why it would be missing in CI in this case. Maybe we should punt?

File: azure-pipelines.yml [hunk 2]
```diff
@@ -242,4 +242,4 @@ jobs:
     - template: eng/pipelines/publish-logs.yml
       parameters:
         jobName: Correctness_Rebuild
-        configuration: Debug
+        configuration: Release
```
Comment (jaredpar) [inline azure-pipelines.yml:245]: Think this will Release the information you were looking for.

**Comment by RikkiGibson** [comment]
Filed #51598 to track the failing rebuilds

File: src/Tools/BuildValidator/LocalReferenceResolver.cs [hunk 3]
```diff
@@ -68,23 +68,36 @@ public string GetReferencePath(MetadataReferenceInfo referenceInfo)
...
+            }
 
-            var metadataReferences = files.Select(f => MetadataReference.CreateFromFile(f)).Cast<MetadataReference>().ToImmutableArray();
-            return metadataReferences;
+            results = builder.MoveToImmutable();
+            return true;
         }
 
-        public void CacheNames(ImmutableArray<MetadataReferenceInfo> names)
+        public bool CacheNames(ImmutableArray<MetadataReferenceInfo> references)
```
Comment (RikkiGibson) [inline src/Tools/BuildValidator/LocalReferenceResolver.cs:95]: incidentally: this is pretty much exactly what PrepareTests does--go searching for a bunch of assemblies, read in their MVIDs, and collect them in a dictionary. Perf improvements are probably possible by using a similar parallelized query like what we did in there.

File: src/Tools/BuildValidator/LocalReferenceResolver.cs [hunk 4]
```diff
@@ -93,40 +106,46 @@ public void CacheNames(ImmutableArray<MetadataReferenceInfo> names)
...
-                    if (matchedReference.FileInfo is null)
+                    foreach (var reference in references)
                     {
-                        continue;
+                        if (reference.FileInfo.Name != file.Name)
+                        {
+                            continue;
+                        }
+
+                        if (GetMvidForFile(file) is not { } mvid)
```
Comment (jaredpar) [inline src/Tools/BuildValidator/LocalReferenceResolver.cs:116]: In the case `GetMvidForFile` is `null` we should probably log a warning.

File: azure-pipelines.yml [hunk 5]
```diff
@@ -234,9 +234,9 @@ jobs:
       displayName: Restore
       inputs:
         filePath: eng/build.ps1
-        arguments: -configuration Debug -prepareMachine -ci -restore -binaryLog
+        arguments: -configuration Debug -prepareMachine -ci -useGlobalNuGetCache:$false -restore -binaryLog
 
-    - powershell: .\eng\test-rebuild.ps1 -ci -configuration Release
+    - powershell: .\eng\test-rebuild.ps1 -ci -useGlobalNuGetCache:$false -configuration Release
```
Comment (jaredpar) [inline azure-pipelines.yml:239]: Do you still need this now that you've removed the problematic reference?

File: src/Tools/BuildValidator/LocalReferenceResolver.cs (hunk 4 above)
Comment (RikkiGibson) [reply src/Tools/BuildValidator/LocalReferenceResolver.cs:116]: Yeah if the filename already matches the reference then it seems bad if we can't read an mvid from it.

File: azure-pipelines.yml (hunk 5 above)
Comment (RikkiGibson) [reply azure-pipelines.yml]: Good catch. I'll remove the arguments but I'll leave the parameter in the test-rebuild script alone for now.

**Comment by RikkiGibson** [comment]
@jaredpar do you want to sign off again or are we good to just merge here?

**Comment by jaredpar** [comment]
Good to merge

**Pull request merged**